### ISSUE TITLE
Original File Creation Date Modified After Sync

### ISSUE THREAD

**Issue by kennygokh** [syncthing/syncthing#9191]
### Does your log mention database corruption?

No

### Include required information

Please be sure to include at least:

 - which version of Syncthing and what operating system you are using 
 v1.25.0, macOS (64-bit ARM)

 - browser and version, if applicable
Chrome, Arc

 - what happened,
A file was created on 03-Nov-23, see screenshot highlighted in red box. I did a simple update to this file, and you can see modified date shown today date & time (05-Nov-23 11:45). However, when I sync to iPhone, the creation date become today's date. 

MacOS Screenshot:
![image](https://github.com/syncthing/syncthing/assets/97268671/7aa7beb3-f22f-41d5-82ec-ca55fe93fb47)

iPhone Screenshot:
![image](https://github.com/syncthing/syncthing/assets/97268671/218748cf-b493-4a9f-86ac-8d225c4a6543)

 - what you expected to happen instead, and
Original file creation date should not be modified. It preserves the creation date of the file which is important for checking subsequently.

 - any steps to reproduce the problem.
Modify the file in macbook, sync it in mobile.

**Comment by acolomb** [comment]
What are you using to sync to the iPhone? Syncthing is not available there.

And are you talking about creation date or modification date? Better to discuss this on the forum, as long as a bug is not clearly identified.

**Comment by kennygokh** [comment]
I am using Mobius sync to sync. I am not sure whether it is due to Mobius sync as it might be caused by the sync from macOS syncthing?

**Comment by tomasz1986** [comment]
The iPhone screenshot only shows one time with no further information. Are you sure that it's supposed to be creation time and not modification time? Usually, when a file explorer only shows one time, it's modification time.

**Comment by kennygokh** [comment]
Yes. See details screenshot. Both dates were modified as the same date & time.
![Uploading Screenshot of 20231103 Fri.jpeg…]()

**Comment by acolomb** [comment]
You will have to take this up with the Möbius Sync people. It's not our product and Syncthing earns nothing from their commercial offering. Thus free support for a commercial software is very limited.

**Comment by kennygokh** [comment]
Ok. I have posted the same at Mobius. By the way, is there a way to use syncthing direct in iOS platform?

**Comment by calmh** [comment]
The iOS/Mobius stuff is a red herring; Syncthing doesn't sync *creation* dates on any platform. Those aren't part of the usual file system APIs and many file systems don't have them to begin with. Nonetheless we could investigate what it would take to do so.

**Comment by tomasz1986** [comment]
> Syncthing doesn't sync _creation_ dates on any platform.

I think this issue may not be strictly about syncing creation time. For example, on Windows, even if I update a file on another device and let Syncthing sync the changes to the current device, the file still retains its original creation time after the update. Here, however, it seems that the creation time gets reset on update.

**Comment by calmh** [comment]
I don't know what you're seeing, but the behaviour described here makes sense as every synced file is created from scratch. It's just that not that many people look at or care about creation timestamps.

**Comment by tomasz1986** [comment]
Normally, at least on Windows (and on macOS too, judging by the first screenshot from this very issue), the original creation time gets retained even when files get updated (either locally or through Syncthing).

**Comment by calmh** [comment]
Syncthing doesn't "update" files, except in the limited circumstance of just changing permissions or other metadata. If the content has changed it's a new file.

**Comment by tomasz1986** [comment]
But "updated" is specifically the term Syncthing itself uses! In the GUI, it's either "updated" or "deleted", never "created" (related issue: https://github.com/syncthing/syncthing/issues/7283). Nevertheless, the point is that the original creation date is normally retained even when modifying the file on another device and syncing those modification to the current device, e.g.

![image](https://github.com/syncthing/syncthing/assets/5626656/d7353759-5c2b-4cc8-9abe-c3e3fb44c3d7)

**Comment by calmh** [comment]
I can pretty much guarantee that it isn't.

**Comment by tomasz1986** [comment]
It is on Windows for sure. Of course, this may be due to how the OS deals with file replacement, not necessarily thanks to Syncthing itself. On the other hand, Android doesn't seem to even support creation time at all. I can't check other OSs right now.

**Comment by kennygokh** [comment]
When the file first created, it should retain the state of the file regardless of which OS platform. That is why there is an attribute like Creation Date. When sync happen, if the file content differs, it should retain the original state of file creation date & time, however the content differ, sync should update only modify date & time. It's the same behavior in other platform like iCloud, OneDrive etc.

**Comment by tomasz1986** [comment]
Related: https://forum.syncthing.net/t/feature-request-sync-creation-times-dates/9184

@kennygokh Can you test what happens the other way around? Basically, just modify the file on the iOS side and see what happens with the creation and modification dates on the Mac.

**Comment by er-pa** [comment]
> It is on Windows for sure. Of course, this may be due to how the OS deals with file replacement, not necessarily thanks to Syncthing itself. On the other hand, Android doesn't seem to even support creation time at all. I can't check other OSs right now.

On Windows it just retains the creation date of when the file is being created on that Windows machine, it still doesn't retain any 'original' creation date from the sending side. But even so, Windows seems to be a special case as MacOS and Linux never retain that creation date.

And indeed, Android with its emulated FS is a special case. There does seem to be some support for it since API 26, but it doesn't seem to result in much.

**Comment by kennygokh** [comment]
> Related: https://forum.syncthing.net/t/feature-request-sync-creation-times-dates/9184
> 
> 
> 
> @kennygokh Can you test what happens the other way around? Basically, just modify the file on the iOS side and see what happens with the creation and modification dates on the Mac.

I am not able to try as I get insufficient disk space error. The same error (see screenshot) encountered even though I created a test vault with a small file. ![image](https://github.com/syncthing/syncthing/assets/97268671/78556732-91b8-4729-a755-bababe31829e)

**Comment by tezlm** [comment]
This is also an issue on linux. Some filesystems like btrfs have birthtime, but syncthing doesn't seem to set it on synced files.

**Comment by calmh** [comment]
I don't think there's even an API to set that timestamp. It cannot be synced.

**Comment by rasa** [comment]
OpenWRT 23.05 doesn't appear to support birthdates in a default install:
```
find / -print0 2>/dev/null | xargs -0 -n 1 stat 2>/dev/null | grep "^Birth:" | grep -v "^Birth: -" | wc
        0         0         0
```

**Comment by keke1229488344** [comment]
Just wanna say as a Windows user, I really need this feature. Windows Explorer enables sorting by folder/file creation time by default, which is so necessary for me to find my documents. So, strongly hoping to this :-)

A temporary solution may be (before this function is implemented), we can use some kind of programming to synchronize the file creation time during the initial synchronization, and then keep syncthing turned on, and it will continue to synchronize new additions files (perhaps with a slight time offset).

**Comment by milroneth** [comment]
To clear some things up for everyone:

The Creation timestamp does not "belong" to the file. It belongs to the filesystem (partition, disk) that contains it. Some filesystems don't even store this at all, so it's not a universal idea. While the modification time _should_ follow the file around, the creation time does not (and it shouldn't, this isn't what it's meant for).

Even in Windows that we are used to rely on Creation Time, it's not guaranteed to be preserved. For example, if you copy a file you made yesterday, you will notice that the copy has a creation time of NOW even if it says modified yesterday. If you then delete the original and keep the copy, it would be almost equivalent as if you moved it, but the original creation time is lost, it is a newly created file after all.

If you move the file around in the same partition (say you make a subfolder and put it in), it's preserved. If you move the file to a USB disk, you will again see that the creation time is lost and changed to now. That is correct, the file was JUST NOW created on the USB stick, even if it was modified on some other disk earlier.

To re-iterate, the creation timestamp is the storage medium telling you when it saw that file for the first time, not when **_you_** first made it. Those 2 just happen to initially coincide when you make a new file from scratch.


So, when you copy or backup files, the creation time is always "new" for the copy, if it even exists at all, as some don't store that value. Modification times should always be preserved and when they are not that is wrong and shouldn't happen, but not so for creation times.

Lastly, as some of you noticed and in some cases, updating a file can also update its creation time to now, along with its modification time. That happens because certain apps don't edit the original file when they have to update it but instead save a whole new file to replace the old one. This is either done unnecessarily, out of weirdness, or for safety reasons. In the later case the new file is initially saved with a new filename, then the old one deleted and the new one renamed to take its place. This saves your file in case the operation is interrupted as the original is still untouched until the new one is safely stored. This is very common behaviour for backup/sync software so that you don't end up with only half-transfered files and no originals and lose your data.

In conclusion, Creation time is useful, but only to an extent, don't count on it. If you need to know when a file was created, put that in the filename.

**Comment by lynrayy** [comment]
Is it fixed?
I want to sync photos FROM my phone TO windows with original file creation dates. Wtf? How to change it?

**Comment by milroneth** [comment]
> Is it fixed? I want to sync photos FROM my phone TO windows with original file creation dates. Wtf? How to change it?

It can't be fixed because it's not broken. Please read my extensive answer that you skipped, it's right above your comment, it will explain everything. Short answer: nothing preserves that, not even Windows itself when you transfer photos from an SD card with the file manager.

**Comment by github-account1111** [comment]
> if you copy a file you made yesterday, you will notice that the copy has a creation time of NOW even if it says modified yesterday. If you then delete the original and keep the copy, it would be almost equivalent as if you moved it, but the original creation time is lost, it is a newly created file after all.

This is only the case for some software but not other. File Explorer, for example, behaves like you're describing. Robocopy, for example, doesn't. It preserves timestamps for both moves and copies by default. It can even do so for directories through flags.

In other words, creation date is guaranteed to be preserved if you copy things correctly.

> If you move the file around in the same partition (say you make a subfolder and put it in), it's preserved. If you move the file to a USB disk, you will again see that the creation time is lost and changed to now. That is correct, the file was JUST NOW created on the USB stick, even if it was modified on some other disk earlier.

Just tried it, and that's not the case. If you really _move_ the file to another disk (instead of copying which File Explorer defaults to when drag-and-dropping to another disk, so you need to press Shift before dropping) creation time remains unchanged.

> To re-iterate, the creation timestamp is the storage medium telling you when it saw that file for the first time, not when you first made it. Those 2 just happen to initially coincide when you make a new file from scratch.

This is not true. As stated above, it entirely depends on the tooling used. Creation time is not inherently dependent on "when the medium saw it the first time." Even Microsoft don't describe it as such anywhere. You're basing your assumptions entirely on a single piece of software (File Explorer). Creation time can very easily be "when you first made it" and is a much more useful piece of information for most people for hopefully obvious reasons.

> f you need to know when a file was created, put that in the filename.

Metadata, including creation date, exists for a reason. I want to use _it_ instead of coming up with ugly, jerry-rigged solutions, like prepending dates to filenames.

> I'm not sure the iPhone even stores creation times at all.

It does. They're on the screenshot.

**Comment by OverloadUT** [comment]
All of these objections about Creation Date not being universal across filesystems or doesn't "belong" to the file is silly, when Syncthing already supports several features that are filesystem- and OS-dependent, like `junctionsAsDirs` (affects NTFS only), `syncOwnership` (totally different data on Unix-like vs Windows), `syncXattrs` (different data per OS). Adding support for Creation Date is well within the spirit of the software.

I'm adding my strong vote for wanting this feature. Obsidian, an open-source tool very much in a similar spirit to Syncthing, is a good example of an application where the file Creation Date is critical metadata.

**Comment by krokosik** [comment]
I second that Obsidian goes amazingly well with Syncthing apart from this issue which greatly messes up date based dynamic vaults. Also, it's not open source ;)

**Comment by 23floors** [comment]
gonna summarize my (shaky) understanding - and workaround - hope that it helps

in my set-up, syncthing syncs between android and my windows pc

all sorts of files are sync'd - photos, downloads, pdfs, obsidian files, etc   (HF thank you syncthing!)

because syncthing creates 'tmp' files for sync'd information - and then copies 'those tmp files' to the 'original files' - 'file creation date' on the 'original files' gets updated

however the 'file name' doesn't change - so this timestamp issue can be a bit unnerving when first noticed

i have files which show as having been modified well before they were actually created

and one would perhaps assume that windows 'knows' how to deal with this - it doesn't - those are just fields in a filesystem database (of sorts) - and it doesn't have to be logical or coherent between create and modify - that isn't it's job - it assumes that you're being coherent - instead you're being wild and using syncthing to remotely sync files with a different operating system

now, many file 'types' contain 'metadata' - and this metadata can usually be read by the program which uses that file (among others) - you can often still see 'when a file was created' depending on where you look

- windows file explorer --> nope
- file through source program --> very good odds - usually something like "file > properties"

when it comes to obsidian, it (mostly) uses .md files

.md files are just .txt files

.md add markdown (presentation) syntax - this is baked into the file 'content', not into the file 'metadata'

and .txt files do NOT contain metadata - they only report what is available through the windows filesystem

so, that's how you can get those funky timestamps

if you're using obsidian, the solution is trivial - (and i assume that there are similar solutions for other platforms)

'hard code' the 'created date' in the YAML header (or even in the message body - there are nuances between the approaches, and YAML is the safer bet, as i recall)

you can 'automate' adding this field using the templater plugin - (many great intros on youtube) 
 
- add to all files - by default
- add when triggered by shortcut / command

then, within obsidian, you can search for that 'created date' value and use it to 'grab / organize' your files

afaik, all of the query methods (search field, query, dataview, bases, etc) will let you search for a field that is set in the YAML header

(thank you bossidian !)

such 'organization' will not work through windows explorer alone though - it reads the windows timestamps

files *can* be further organized, if you parse the file contents with scripting (must have metadata contents to parse - one of the benefits of YAML) - that definitely can be done

---

perhaps syncthing could save a target file's 'create date', then add it back after the copy operation - i dunno

IF YOU NEED windows filesystem timestamps to be preserved for whatever reason - you may want to consider a different solution, because this 'create date' stuff appears to be fundamental to how syncthing and windows interact

---

oh - and if you delete your syncthing database or deltas (either on the phone or pc) because you're troubleshooting, you 'may' have aggravated this situation

i definitely do not understand the role of these files vs the windows timestamps that are printed (!)

---

please chime in with any bonifications - thanks!

**Comment by ienwb** [comment]
Yeah Obsidian is not open source, but its plugin community, tools, and stance is very open for community developments, not stagnant and non-progressive.

I never change sync conflict in Obsidian anymore. Just gotta flow with the latest file, changing it only delete then instead of only deleting one of the two.

Obsidian, KeePass fork are 2 softwares where they really need correct creation date, modified date; but more important in Obsidian. Linter (plugin) also doesnt work really well in Obsidian probably the differences of linter, OS-dependent, and syncthing timestamps.
